src/Rules/Documentation.hs

```diff
@@ -36,12 +119,72 @@ buildPackageDocumentation context@Context {..} = when (stage == Stage1) $ do
         let haddockWay = if dynamicPrograms then dynamic else vanilla
         build $ target (context {way = haddockWay}) Haddock srcs [file]
 
-    when (package == haddock) $ haddockHtmlLib %> \_ -> do
-        let dir = takeDirectory haddockHtmlLib
-        liftIO $ removeFiles dir ["//*"]
-        copyDirectory "utils/haddock/haddock-api/resources/html" dir
+-- | Build the haddocks for GHC's libraries
+buildLibraryDocumentation :: Rules ()
+buildLibraryDocumentation = do
+    "//" ++ htmlRoot -/- "libraries/index.html" %> \file -> do
+        haddocks <- allHaddocks
+        need haddocks
+        let libDocs = filter (\x -> takeFileName x /= "ghc.haddock") haddocks
+        let iface haddock = "--read-interface="
+                            ++ (takeFileName . takeDirectory) haddock ++ ","
+                            ++ haddock
+        runBuilder Haddock ([ "--gen-index"
+                            , "--gen-contents"
+                            , "-o", takeDirectory file
+                            , "-t", "Haskell Hierarchical Libraries"
+                            , "-p", "libraries/prologue.txt" ]
+                            ++ map iface libDocs)
+                           [] [file]
+  where
+    allHaddocks = do
+        pkgs <- stagePackages Stage1
+        sequence [ pkgHaddockFile $ vanillaContext Stage1 pkg
+                 | pkg <- pkgs, isLibrary pkg, isHsPackage pkg
+                                             , pkgName pkg /= "Win32" ]
+
+----------------------------------------------------------------------
+-- PDF
+
+-- | Build all PDF documentation
+buildPdfDocumentation :: Rules ()
+buildPdfDocumentation = mapM_ buildSphinxPdf docPaths
+
+-- | Compile a Sphinx ReStructured Text package to LaTeX
+buildSphinxPdf :: FilePath -> Rules ()
+buildSphinxPdf path = do
+    "//" ++ path <.> "pdf" %> \file -> do
+        let context = vanillaContext Stage0 docPackage
+        withTempDir $ \dir -> do
+            -- TODO: Figure out a crutch
+            build $ target context (Sphinx Latex) [pathPath path] [dir]
+            --runBuilderWithCmdOptions [Cwd dir] Xelatex ["-halt-on-error"
+            --                                           , name <.> "tex" ] [] []
+            unit $ cmd Shell [Cwd dir] ["xelatex", "-halt-on-error"
```

Comment:
I think it should be possible to add a custom build action for `Xelatex` here:

https://github.com/snowleopard/hadrian/blob/master/src/Builder.hs#L149

We can pass the temporary directory via the `outputs` field of the `Target`. For example:
```haskell
            case builder of
                Xelatex -> do
                    cmd Shell [Cwd output] [path] buildArgs
                    cmd Shell [Cwd output] [path] buildArgs
                    cmd Shell [Cwd output] [path] buildArgs
                    cmd Shell [Cwd output] ["makeindex"] (input -<.> "idx")
                    cmd Shell [Cwd output] [path] buildArgs
                    cmd Shell [Cwd output] [path] buildArgs
```
Then this rule becomes:
```haskell
        withTempDir $ \dir -> do
             build $ target context (Sphinx Latex) [pathPath path] [dir]
             build $ target context Xelatex [path <.> "tex"] [dir]
             copyFileUntracked (dir -/- path <.> "pdf") file
```